aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -57,25 +57,24 @@ Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
 }
 
 Tensor logical_not(const Tensor& self) {
-  Tensor result = at::empty({0}, self.options());
+  Tensor result = at::empty({0}, self.options().dtype(kBool));
   return at::logical_not_out(result, self);
 }
 
 Tensor& logical_not_(Tensor& self) {
-  return at::logical_not_out(self, self);
+  auto iter = TensorIterator::unary_op(self, self, /*check_internal_overlap=*/true);
+  logical_not_inplace_stub(iter.device_type(), iter);
+  return self;
 }
 
 Tensor& logical_not_out(Tensor& result, const Tensor& self) {
-  TORCH_CHECK(self.scalar_type() == kBool,
-              "logical_not currently only supports bool tensors.");
-  TORCH_CHECK(result.scalar_type() == kBool,
-              "The output tensor of logical_not must be a bool tensor.");
-  auto iter = TensorIterator::unary_op(result, self,
-    /*check_internal_overlap=*/true);
+  TORCH_CHECK(result.scalar_type() == kBool, "The output tensor of logical_not must be a bool tensor.");
```

Comment:
so, I don't know if TensorIterator currently supports this, but in NumPy I can have any return type (as long it's the same device?) and it will just do the cast for me.  I don't understand why we would need different behavior for in-place vs not.